odoo/tools/js_sourcemap.py

```diff
@@ -0,0 +1,171 @@
+                    result += ";"
+                    previous_generated_line += 1
+
+            if mapping["source"] is not None:
+                sourceIdx = self._sources[mapping["source"]]
+                source = sourceIdx - previous_source
+                previous_source = sourceIdx
+
+                # lines are stored 0-based in SourceMap spec version 3
+                line = mapping["originalLine"] - 1 - previous_original_line
+                previous_original_line = mapping["originalLine"] - 1
+
+            if (source, line) not in self._cache:
+                self._cache[(source, line)] = "".join([
+                    encoded_column,
+                    base64vlq_encode(source),
+                    base64vlq_encode(line),
+                    encoded_column,
+                ])
+
+            result += self._cache[source, line]
+        return result
+
+    def to_json(self):
+        """
+        Generates the json sourcemap.
+        It is the main function that assembles all the pieces.
+
+        :return {str} valid sourcemap in json format
+        """
+        mapping = {
+            "version": self._version,
+            "sources": list(self._sources.keys()),
+            "mappings": self._serialize_mappings(),
+            "sourcesContent": [self._sources_contents[source] for source in self._sources]
+        }
+        if self._file:
+            mapping["file"] = self._file
+
+        if self._source_root:
+            mapping["sourceRoot"] = self._source_root
+
+        return mapping
+
+    def add_source(self, source_name, source_content, last_index, start_offset=0):
+        """Adds a new source file in the sourcemap. All the lines of the source file will be mapped line by line
+        to the generated file from the (last_index + start_offset). All lines between
+        last_index and (last_index + start_offset) will
+        be mapped to line 1 of the source file.
+
+        Example:
+            ls 1 = Line 1 from new source file
+            lg 1 = Line 1 from genereted file
+            ls 1 <=> lg 1 Line 1 from new source file is map to  Line 1 from genereted file
+            nb_ls = number of lines in the new source file
+
+            Step 1:
+            ls 1 <=> lg last_index + 1
+
+            Step 2:
+            ls 1 <=> lg last_index + start_offset + 1
+            ls 2 <=> lg last_index + start_offset + 2
+            ...
+            ls nb_ls <=> lg last_index + start_offset + nb_ls
+
+
+        :param source_name: name of the source to add
+        :param source_content: content of the source to add
+        :param last_index: Line where we start to map the new source
+        :param start_offset: Number of lines to pass in the generated file before starting mapping line by line
+        """
+        source_line_count = len(source_content.split("\n"))
+
+        if (source_name not in self._sources):
+            self._sources[source_name] = len(self._sources)
```

Comment:
more idiomatic way to do this in Python:

```suggestion
        self._sources.setdefault(source_name, len(self._sources))
```